libgin/archive/archive_test.go

```diff
@@ -268,3 +268,137 @@ func TestTar(t *testing.T) {
+		}
+
+		mkfile := func(currfile string) error {
+			fp, err := os.Create(currfile)
+			if err != nil {
+				return fmt.Errorf("Error creating file %s: %v", currfile, err)
+			}
+			fp.Close()
+			return nil
+		}
+		// Create files
+		if err = mkfile(filepath.Join(ziproot, "included.md")); err != nil {
+			return err
+		}
+		if err = mkfile(filepath.Join(gitpath, "excluded.md")); err != nil {
+			return err
+		}
+		if err = mkfile(filepath.Join(inclpath, "included.md")); err != nil {
+			return err
+		}
+		if err = mkfile(filepath.Join(inclpath, "not_excluded.md")); err != nil {
+			return err
+		}
+		if err = mkfile(filepath.Join(exclpath, "excluded.md")); err != nil {
+			return err
+		}
+
+		return nil
+	}
+	if err = handletestdir(); err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	zipbasename := "test_makezip.zip"
+	zipfilename := filepath.Join(targetpath, zipbasename)
+
+	// Checks that files in directories ".git" and ".excluded" are excluded and 
+	// file "not_excluded" is still added.
+	exclude := []string{".git", ".excluded", "not_excluded.md"}
+
+	handlezip := func() error {
+		fn := fmt.Sprintf("zip(%s, %s)", ziproot, zipfilename)
+		source, err := filepath.Abs(ziproot)
+		if err != nil {
+			return fmt.Errorf("Failed to get abs path for source directory in function '%s': %v", fn, err)
+		}
+
+		zipfilename, err = filepath.Abs(zipfilename)
+		if err != nil {
+			return fmt.Errorf("Failed to get abs path for target zip file in function '%s': %v", fn, err)
+		}
+	
+		zipfp, err := os.Create(zipfilename)
+		if err != nil {
+			return fmt.Errorf("Failed to create zip file for writing in function '%s': %v", fn, err)
+		}
+		defer zipfp.Close()
+	
+		origdir, err := os.Getwd()
+		if err != nil {
+			return fmt.Errorf("Failed to get working directory in function '%s': %v", fn, err)
+		}
+		defer os.Chdir(origdir)
+	
+		if err := os.Chdir(source); err != nil {
+			return fmt.Errorf("Failed to change to source directory to make zip file in function '%s': %v", fn, err)
+		}
+	
+		if err := MakeZip(zipfp, exclude, "."); err != nil {
+			return fmt.Errorf("Failed to change to source directory to make zip file in function '%s': %v", fn, err)
+		}
+		return nil
+	}
+
+	if err = handlezip(); err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	// Files included in the zip file
+	incl := map[string]struct{} {
+		"included/included.md": {},
+		"included/not_excluded.md": {},
+		"included.md": {},
+	}
+	// Files excluded from the zip file
+	excl := map[string]struct{} {
+		".git/excluded.md": {},
+		"excluded/excluded.md": {},
+	}
+
+	zipreader, err := zip.OpenReader(zipfilename)
+	if err != nil {
+		t.Fatalf("Error opening zip file: %v", err)
+	}
+	defer zipreader.Close()
+
+	var includedCounter []string
+	for _, file := range zipreader.File {
+		if _, included := incl[file.Name]; !included {
+			if _, notExcluded := excl[file.Name]; notExcluded {
+				t.Fatalf("Unexpected file found: %s", file.Name)
+			}
```

Comment:
This raises an error if a file found in the zip file was on the exclusion list, but shouldn't any unexpected file (i.e., any file not in the `incl` map) raise an error?

Reply:
In an admittedly not too obvious way, the `len` check below accounts for any unexpected files showing up in the zip file.

Since we control the test environment, no files should be present that are added to the zip file but are not mentioned in the `incl` map. As a sanity check the `len` check will catch it if the test setup goes awry and also report a full list which files have been added to the zip to pinpoint which file was unexpectedly added.